# Финальный проект по SQL

**Описание проекта**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задачи:**
- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Описание данных:**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Импортируем библиотеки

In [1]:
import pandas as pd
from sqlalchemy import create_engine

## Прочитаем файл

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Изучим таблицы

### Выведем таблицу books

In [3]:
query = '''
SELECT *
FROM books
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Выведем таблицу authors

In [4]:
query = '''
SELECT *
FROM authors
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### Выведем таблицу publishers

In [5]:
query = '''
SELECT *
FROM publishers
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Выведем таблицу ratings

In [6]:
query = '''
SELECT *
FROM ratings
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### Выведем таблицу reviews

In [7]:
query = '''
SELECT *
FROM reviews
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Данные содержат информацию, представленную в тз, можем приступать к выполнению задания.

## Выполним задание

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [8]:
query = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date >= '2000-01-01'
'''
pd.io.sql.read_sql(query, con = engine)

count
0    821

После 1 января 2000 года вышло 821 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [9]:
query = '''
WITH
books_review_counts AS
  (SELECT book_id,
          COUNT(review_id) AS review_count
   FROM reviews
   GROUP BY book_id),
 
books_avg_ratings AS 
  (SELECT book_id,
          AVG(rating) AS avg_rating
   FROM ratings
   GROUP BY book_id)

SELECT books.book_id,
       books.title,
       books_review_counts.review_count,
       books_avg_ratings.avg_rating
FROM books
LEFT JOIN books_review_counts ON books.book_id = books_review_counts.book_id
LEFT JOIN books_avg_ratings ON books.book_id = books_avg_ratings.book_id
'''

pd.io.sql.read_sql(query, con = engine) 

book_id                                              title  review_count  \
0        652          The Body in the Library (Miss Marple  #3)           2.0   
1        273                                          Galápagos           2.0   
2         51                           A Tree Grows in Brooklyn           5.0   
3        951  Undaunted Courage: The Pioneering First Missio...           2.0   
4        839                                        The Prophet           4.0   
..       ...                                                ...           ...   
995      672   The Cat in the Hat and Other Dr. Seuss Favorites           NaN   
996       83    Anne Rice's The Vampire Lestat: A Graphic Novel           NaN   
997      221                          Essential Tales and Poems           NaN   
998      387                               Leonardo's Notebooks           NaN   
999      808                            The Natural Way to Draw           NaN   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    5.000000  
996    3.666667  
997    4.000000  
998    4.000000  
999    3.000000  

[1000 rows x 4 columns]

Не на все книги пользователи написали отзыв.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [10]:
query = '''
SELECT publishers.publisher,
       COUNT(books.book_id)
FROM books 
LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher_id
ORDER BY COUNT(books.book_id) DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

Больше всего книг толще 50 страниц выпустило издательство "Penguin Books".

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
query = '''
WITH 
a AS (
SELECT authors.author,
       AVG(ratings.rating) AS avg_rating
FROM books 
LEFT JOIN authors ON books.author_id = authors.author_id
LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY authors.author_id, books.book_id
HAVING COUNT(ratings.rating_id) >= 50)

SELECT a.author,
       AVG(a.avg_rating)
FROM a
GROUP BY a.author
ORDER BY AVG(a.avg_rating) DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

Самая высокий показатель среднего рейтинга книг у дуэта J.K. Rowling/Mary GrandPré.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''
WITH
users AS (
SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(rating_id) > 50
),

rev AS (
SELECT COUNT(review_id) AS rev_id
FROM reviews
WHERE username IN (SELECT *
                   FROM users)
GROUP BY username
) 

SELECT AVG(rev_id)
FROM rev
'''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

Пользователи, которые поставили больше 50 оценок в среднем пишут по 24 отзыва.

## Общий вывод

- После 1 января 2000 года вышло 821 книг.
- Не на все книги пользователи написали отзыв.
- Больше всего книг толще 50 страниц выпустило издательство "Penguin Books".
- Самая высокий показатель среднего рейтинга книг у дуэта J.K. Rowling/Mary GrandPré.
- Пользователи, которые поставили больше 50 оценок в среднем пишут по 24 отзыва.

Одним из наиболее удачных вариантов для нового продукты - это книги издания "Penguin Books", автором, которых является J.K. Rowling/Mary GrandPré.